### ISSUE TITLE
File matchers should implicitly check File.shouldExist

### ISSUE THREAD

**Issue by sgarfinkel** [kotest/kotest#764]
This should either be a documented feature (you need to always check `shouldExist` before calling any other matcher file matcher) or should just be done implicitly for the user (my preferred approach). Because the matchers just defer to the Java File/Path checks, they have no null checking. This results in unexpected NPEs.

A `shouldExist` check, in my mind, is implied if you are using any of the other matchers. In my use case, this was resulting from a call like:

```kotlin
File("foo").resolve("bar") shouldContainFile "xy.z"
```

Where the subdirectory "bar" did not exist.

Additionally, I think the message of the `shouldExist` matcher could be improved to highlight the first intermediate path that does not exist. That probably belongs in another issue, though.

**Comment by sksamuel** [comment]
Without checking myself, are you saying that File("qeqweqweqweqwewqe").resolve("bar") throws an NPE ?

**Comment by sgarfinkel** [comment]
@sksamuel No, its thrown from the matcher. Resolve just returns a File, which may or may not actually exist.

The specific issue in the example above is that `shouldContainFile` eventually calls `contents = File.list()` which returns null when File is either not a directory or does not exist. The returned result then tries to call `contents.size` which causes the NPE.

Here's the full function for reference: 
```kotlin
fun containFile(name: String) = object : Matcher<File> {
  override fun test(value: File): Result {
    val contents = value.list()
    val passed = value.isDirectory && contents.contains(name)
    return Result(passed,
        "Directory $value should contain a file with filename $name (detected ${contents.size} other files)",
        "Directory $value should not contain a file with filename $name"
    )
  }
}
```

**Comment by sksamuel** [comment]
Ah yes, .list() can return null.
Good call thanks. These matchers should definitely handle files that don't exist.
Your 2nd point is good too - the intermediate path, very useful. I'll cc in the matchers expert @Kerooker

**Comment by sksamuel** [comment]
Fixed in master for 4.0 release.
